File: src/main/java/net/sf/rails/game/MapManager.java
package net.sf.rails.game;

import java.util.*;
import java.util.Map.Entry;

import net.sf.rails.common.Config;
import net.sf.rails.common.parser.Configurable;
import net.sf.rails.common.parser.ConfigurationException;
import net.sf.rails.common.parser.Tag;

import com.google.common.collect.HashBasedTable;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSortedSet;
import com.google.common.collect.ImmutableTable;
import com.google.common.collect.Table;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * MapManager configures the map layout from XML
 */
public class MapManager extends RailsManager implements Configurable {

    private MapOrientation mapOrientation;

    private ImmutableMap<MapHex.Coordinates, MapHex> hexes;
    private ImmutableTable<MapHex, HexSide, MapHex> hexTable;

    private MapHex.Coordinates minimum;
    private MapHex.Coordinates maximum;

    // upgrade costs on the map for noMapMode
    private ImmutableSortedSet<Integer> possibleTileCosts;

    // Stop property defaults per stop type
    private EnumMap<Stop.Type, Access> defaultAccessTypes = new EnumMap<>(Stop.Type.class);
    
    // if required: distance table
    private Table<MapHex, MapHex, Integer> hexDistances;

    // Optional map image (SVG file)
    // FIXME: Move to UI class
    private String mapImageFilename = null;
    private String mapImageFilepath = null;
    private int mapXOffset = 0;
    private int mapYOffset = 0;
    private float mapScale = (float)1.0;
    private boolean mapImageUsed = false;

    private static final Logger log = LoggerFactory.getLogger(MapManager.class);

    /**
     * Used by Configure (via reflection) only
     */
    public MapManager(RailsRoot parent, String id) {
        super(parent, id);
    }
    
    /*
     * @see net.sf.rails.common.parser.Configurable#configureFromXML(org.w3c.dom.Element)
     */
    public void configureFromXML(Tag tag) throws ConfigurationException {
        
        mapOrientation = MapOrientation.create(tag);

        List<Tag> hexTags = tag.getChildren("Hex");
        ImmutableMap.Builder<MapHex.Coordinates, MapHex> hexBuilder = ImmutableMap.builder();
        ImmutableSortedSet.Builder<Integer> tileCostsBuilder= ImmutableSortedSet.naturalOrder();

        for (Tag hexTag : hexTags) {
            MapHex hex = MapHex.create(this, hexTag);
            hexBuilder.put(hex.getCoordinates(), hex);
            tileCostsBuilder.addAll(hex.getTileCostsList());
        }
        hexes = hexBuilder.build();
        possibleTileCosts = tileCostsBuilder.build();
        
        minimum = MapHex.Coordinates.minimum(hexes.values());
        maximum = MapHex.Coordinates.maximum(hexes.values());
        
        // Default Stop Types
        Tag defaultsTag = tag.getChild("Defaults");
        if (defaultsTag != null) {
            List<Tag> accessTags = defaultsTag.getChildren("Access");
            defaultAccessTypes = Access.parseDefaults(this, accessTags);
        }

        // Map image attributes
        // FIXME: Move to an UI class
        Tag mapImageTag = tag.getChild("Image");
        if (mapImageTag != null) {
            mapImageFilename = mapImageTag.getAttributeAsString("file");
            mapXOffset = mapImageTag.getAttributeAsInteger("x", mapXOffset);
            mapYOffset = mapImageTag.getAttributeAsInteger("y", mapYOffset);
            mapScale = mapImageTag.getAttributeAsFloat("scale", mapScale);
        }
    }

    public void finishConfiguration (RailsRoot root) throws ConfigurationException {

        for (MapHex hex:hexes.values()) {
            hex.finishConfiguration(root);
        }

        // Initialise the neighbours
        ImmutableTable.Builder<MapHex, HexSide, MapHex> hexTableBuilder = ImmutableTable.builder();
        for (MapHex hex:hexes.values()) {
           for (HexSide side:HexSide.all()){
                MapHex neighbour = hexes.get(mapOrientation.
                        getAdjacentCoordinates(hex.getCoordinates(), side));
                if (neighbour != null) {
                    if (hex.isValidNeighbour(neighbour, side)) {
                        hexTableBuilder.put(hex, side, neighbour);
                        if (hex.isRiverNeighbour(neighbour)) {
                            hex.addRiverSide(side);
                            neighbour.addRiverSide(side.opposite());
                        }
                    } else {
                        hex.addInvalidSide(side);
                        if (hex.isImpassableNeighbour(neighbour)) {
                            hex.addImpassableSide(side);
                            neighbour.addImpassableSide(side.opposite());
                        }
                    }
                } else { // neighbour is null
                    hex.addInvalidSide(side);
                }
            }
        }
        hexTable = hexTableBuilder.build();

        for (PublicCompany company : root.getCompanyManager().getAllPublicCompanies()) {
            List<MapHex> homeHexes = company.getHomeHexes();
            if (homeHexes != null) {
                for (MapHex homeHex : homeHexes) {
                   int homeNumber = company.getHomeCityNumber();
                   Stop home = homeHex.getRelatedStop(homeNumber);
                   if (home == null && homeNumber != 0) {
                       throw new ConfigurationException ("Invalid home number "+homeNumber+" for hex "+homeHex
                               +" which has "+homeHex.getStops().size()+" stop");
                   } else {
                       homeHex.addHome(company, home);
                   }
                }
            }
            MapHex hex = company.getDestinationHex();
            if (hex != null) {
                hex.addDestination(company);
            }
        }

        // FIXME: Move this configuration to an UI class
        mapImageUsed = net.sf.rails.util.Util.hasValue(mapImageFilename)
        && "yes".equalsIgnoreCase(Config.get("map.image.display"));
        if (mapImageUsed) {
            String rootDirectory = Config.get("map.root_directory");
            if (!net.sf.rails.util.Util.hasValue(rootDirectory)) {
                rootDirectory = "data";
            }
            mapImageFilepath = "/" + rootDirectory + "/" + mapImageFilename;
        }

    }

    /**
     * @return Returns the currentTileOrientation.
     */
    public MapOrientation getMapOrientation() {
        return mapOrientation;
    }

    /**
     * @return Returns the hexes.
     */
    public Collection<MapHex> getHexes() {
        return hexes.values();
    }
    
    
    public MapHex getNeighbour(MapHex hex, HexSide side) {
        return hexTable.get(hex, side);
    }

    /**
     * A utility to find if a newly laid tile will create a "bridge"
     * (i.e. connect track across a "river" between neighbouring hexes).
     * It is placed here because it is used in both the UI and the server.
     *
     * @param hex The MapHex where a new tile is (being) laid.
     * @param newTile The tile (being) laid.
     * @param newTileRotation The rotation of that tile.
     * @return A HexSideSet object that describes any new bridges,
     * or null if there isn't any.
     */
    public HexSidesSet findNewBridgeSides(MapHex hex, Tile newTile, int newTileRotation) {

        // Has this hex any river sides?
        HexSidesSet rivers = hex.getRiverSides();
        logHexSides(rivers, "River:");
        if (rivers.isEmpty()) return null;

        // Which tracks has the newly laid tile?
        HexSidesSet newTracks = hex.getTrackSides(newTile, newTileRotation);
        logHexSides(newTracks, "New tile track:");

        // Which tracks has the current (i.e. old) tile?
        HexSidesSet oldTracks = hex.getTrackSides();
        logHexSides(oldTracks, "Old tile track:");

        // Which tracks are really new?
        newTracks = newTracks.symDiff(oldTracks);
        logHexSides(newTracks, "New track:");

        // Which new tracks reach a river?
        HexSidesSet bridgePoints = rivers.intersection(newTracks);
        logHexSides(bridgePoints, "New track+river:");
        if (bridgePoints.isEmpty()) return null;

        // Do neighbours have track reaching the same river sides?
        BitSet newBridgeBitSet = new BitSet();
        Iterator it = bridgePoints.iterator();
        while (it.hasNext()) {
            HexSide side = (HexSide) it.next();
            MapHex neighbour = getNeighbour(hex, side);
            Tile nbTile = neighbour.getCurrentTile();
            int nbRot = neighbour.getCurrentTileRotation().getTrackPointNumber();
            if (neighbour.getTrackSides(nbTile, nbRot).get(side.opposite())) {
                newBridgeBitSet.set(side.getTrackPointNumber(), true);
            }
        }
        HexSidesSet newBridges = HexSidesSet.create(newBridgeBitSet);
        logHexSides(newBridges, "New bridge:");

        return newBridges.isEmpty() ? null : newBridges;
    }

    // For debugging
    public void logHexSides(HexSidesSet sides, String prefix) {
        if (!log.isDebugEnabled() || sides == null) return;
        Iterator<HexSide> it = sides.iterator();
        while (it.hasNext()) {
            log.debug("{} {}", prefix, it.next());
        }
    }

    public MapHex getHex(String locationCode) {
        // MapManager is a RailsManager so it is possible to locate by id
        return (MapHex) locate(locationCode);
    }

    public MapHex.Coordinates getMinimum() {
        return minimum;
    }
    
    public MapHex.Coordinates getMaximum() {
        return maximum;
    }
    
    public String getMapUIClassName() {
        return mapOrientation.getUIClassName();
    }

    public Access getDefaultAccessType(Stop.Type type) {
        return defaultAccessTypes.get(type);
    }

    public List<Stop> getCurrentStops() {
        ImmutableList.Builder<Stop> stops = ImmutableList.builder();
        for (MapHex hex : hexes.values()) {
            stops.addAll(hex.getStops());
        }
        return stops.build();
    }

    public SortedSet<Integer> getPossibleTileCosts() {
        return possibleTileCosts;
    }

    public List<MapHex> parseLocations (String locationCodes)
            throws IllegalArgumentException {

        ImmutableList.Builder<MapHex> locationBuilder = ImmutableList.builder(); 
        for (String hexName : locationCodes.split(",")) {
            MapHex hex = getHex(hexName);
            if (hex != null) {
                locationBuilder.add(hex);
            } else {
                throw new IllegalArgumentException("Invalid hex "+hexName+
                        " specified in location string " + locationCodes);
            }
        }
        return locationBuilder.build();
    }

    /**
     * Calculate the distance between two hexes as in 1835,
     * i.e. as "the crow without a passport flies".
     */
    public int getHexDistance (MapHex hex1, MapHex hex2) {
        if (hexDistances == null) {
            hexDistances = HashBasedTable.create();
        }

        if (!hexDistances.contains(hex1, hex2)) {
            calculateHexDistances(hex1, hex1, 0);
        }
        return hexDistances.get(hex1, hex2);
    }

    private void calculateHexDistances (MapHex initHex, MapHex currentHex, int depth) {
        hexDistances.put(initHex, currentHex, depth);
        
        // check for next hexes
        depth ++;
        for (MapHex nextHex:hexTable.row(currentHex).values()) {
            if (!hexDistances.contains(initHex, nextHex) ||
                    depth < hexDistances.get(initHex, nextHex)) {
                calculateHexDistances(initHex, nextHex, depth);
            }
        }
    }


    /**
     * Calculate the distances between a given tokenable city hex
     * and all other tokenable city hexes.
     * <p> Distances are cached.
     * @param initHex Start hex
     * @return Sorted integer list containing all occurring distances only once.
     */
    public SortedSet<Integer> getCityDistances (MapHex initHex) {
        
        if (hexDistances == null) {
            hexDistances = HashBasedTable.create();
        }

        if (!hexDistances.containsRow(initHex)) {
            calculateHexDistances(initHex, initHex, 0);
        }
        
        ImmutableSortedSet.Builder<Integer> distances = 
                ImmutableSortedSet.naturalOrder();
        
        for (Entry<MapHex, Integer> otherHex:hexDistances.row(initHex).entrySet()) {
            if (otherHex.getKey().getCurrentTile().hasStations()) {
                distances.add(otherHex.getValue());
            }
        }
        return distances.build();
    }

    public String getMapImageFilepath() {
        return mapImageFilepath;
    }

    public int getMapXOffset() {
        return mapXOffset;
    }

    public int getMapYOffset() {
        return mapYOffset;
    }

    public float getMapScale() {
        return mapScale;
    }

    public void setMapXOffset(int mapXOffset) {
        this.mapXOffset = mapXOffset;
    }

    public void setMapYOffset(int mapYOffset) {
        this.mapYOffset = mapYOffset;
    }

    public void setMapScale(float mapScale) {
        this.mapScale = mapScale;
    }

    public boolean isMapImageUsed() {
        return mapImageUsed;
    }

}

File: dotnet/RailsLib.Net/Net/Game/MapManager.cs
using GameLib.Net.Common;
using GameLib.Net.Common.Parser;
using System;
using System.Collections.Generic;
using System.Text;

/**
 * MapManager configures the map layout from XML
 */

namespace GameLib.Net.Game
{
    public class MapManager : RailsManager, IConfigurable
    {
        private MapOrientation mapOrientation;

        private Dictionary<MapHex.Coordinates, MapHex> hexes;
        //private ImmutableTable<MapHex, HexSide, MapHex> hexTable;
        private Table<MapHex, HexSide, MapHex> hexTable;

        private MapHex.Coordinates minimum;
        private MapHex.Coordinates maximum;

        // upgrade costs on the map for noMapMode
        private List<int> possibleTileCosts; // Note, this will be sorted

        // Stop property defaults per stop type
        private IReadOnlyDictionary<string, StopType> defaultStopTypes;

        // if required: distance table
        private Table<MapHex, MapHex, int> hexDistances;

        // Optional map image (SVG file)
        // FIXME: Move to UI class
        private string mapImageFilename = null;
        private string mapImageFilepath = null;
        private int mapXOffset = 0;
        private int mapYOffset = 0;
        private float mapScale = (float)1.0;
        private bool mapImageUsed = false;

        /**
         * Used by Configure (via reflection) only
         */
        public MapManager(RailsRoot parent, string id) : base(parent, id)
        {

        }

        /**
         * @see net.sf.rails.common.parser.Configurable#configureFromXML(org.w3c.dom.Element)
         */
        public void ConfigureFromXML(Tag tag)
        {
            mapOrientation = MapOrientation.Create(tag);

            List<Tag> hexTags = tag.GetChildren("Hex");
            Dictionary<MapHex.Coordinates, MapHex> hexBuilder = new Dictionary<MapHex.Coordinates, MapHex>();
            List<int> tileCostsBuilder = new List<int>();

            foreach (Tag hexTag in hexTags)
            {
                MapHex hex = MapHex.Create(this, hexTag);
                hexBuilder[hex.GetCoordinates()] = hex;
                tileCostsBuilder.AddRange(hex.TileCostsList);
            }
            hexes = hexBuilder;
            tileCostsBuilder.Sort();
            possibleTileCosts = tileCostsBuilder;

            minimum = MapHex.Coordinates.Minimum(hexes.Values);
            maximum = MapHex.Coordinates.Maximum(hexes.Values);

            // Default Stop Types
            Tag defaultsTag = tag.GetChild("Defaults");
            if (defaultsTag != null)
            {
                List<Tag> accessTags = defaultsTag.GetChildren("Access");
                defaultStopTypes = StopType.ParseDefaults(this, accessTags);
            }
            else
            {
                defaultStopTypes = new Dictionary<string, StopType>();
            }

            // Map image attributes
            // FIXME: Move to an UI class
            Tag mapImageTag = tag.GetChild("Image");
            if (mapImageTag != null)
            {
                mapImageFilename = mapImageTag.GetAttributeAsString("file");
                mapXOffset = mapImageTag.GetAttributeAsInteger("x", mapXOffset);
                mapYOffset = mapImageTag.GetAttributeAsInteger("y", mapYOffset);
                mapScale = mapImageTag.GetAttributeAsFloat("scale", mapScale);
            }
        }

        public void FinishConfiguration(RailsRoot root)
        {
            foreach (MapHex hex in hexes.Values)
            {
                hex.FinishConfiguration(root);
            }

            // Initialize the neighbors
            Table<MapHex, HexSide, MapHex> hexTableBuilder = new Table<MapHex, HexSide, MapHex>();
            foreach (MapHex hex in hexes.Values)
            {
                foreach (HexSide side in HexSide.All())
                {
                    var h = mapOrientation.GetAdjacentCoordinates(hex.GetCoordinates(), side);
                    MapHex neighbor = null;
                    if (hexes.ContainsKey(h))
                    {
                        neighbor = hexes[h];
                    }
                    if (neighbor != null)
                    {
                        if (hex.IsValidNeighbor(neighbor, side))
                        {
                            hexTableBuilder.Put(hex, side, neighbor);
                        }
                        else
                        {
                            hex.AddInvalidSide(side);
                            if (hex.IsImpassableNeighbor(neighbor))
                            {
                                hex.AddImpassableSide(side);
                                neighbor.AddImpassableSide(side.Opposite);
                            }
                        }
                    }
                    else
                    { // neighbor is null
                        hex.AddInvalidSide(side);
                    }
                }
            }
            hexTable = hexTableBuilder;

            foreach (PublicCompany company in root.CompanyManager.GetAllPublicCompanies())
            {
                List<MapHex> homeHexes = company.GetHomeHexes();
                if (homeHexes != null)
                {
                    foreach (MapHex homeHex in homeHexes)
                    {
                        int homeNumber = company.HomeCityNumber;
                        Stop home = homeHex.GetRelatedStop(homeNumber);
                        if (home == null && homeNumber != 0)
                        {
                            throw new ConfigurationException("Invalid home number " + homeNumber + " for hex " + homeHex
                                    + " which has " + homeHex.Stops.Count + " stop");
                        }
                        else
                        {
                            homeHex.AddHome(company, home);
                        }
                    }
                }
                MapHex hex = company.DestinationHex;
                if (hex != null)
                {
                    hex.AddDestination(company);
                }
            }

            // FIXME: Move this configuration to an UI class
            // #FIXME_GUI_INTERFACE
            mapImageUsed = !string.IsNullOrEmpty(mapImageFilename)
                && "yes".Equals(Config.Get("map.image.display"), StringComparison.OrdinalIgnoreCase);
            if (mapImageUsed)
            {
                string rootDirectory = Config.Get("map.root_directory");
                if (string.IsNullOrEmpty(rootDirectory))
                {
                    rootDirectory = "data";
                }
                mapImageFilepath = "/" + rootDirectory + "/" + mapImageFilename;
            }

        }

        /**
         * @return Returns the currentTileOrientation.
         */
        public MapOrientation MapOrientation
        {
            get
            {
                return mapOrientation;
            }
        }

        /**
         * @return Returns the hexes.
         */
        public ICollection<MapHex> GetHexes()
        {
            return hexes.Values;
        }


        public MapHex GetNeighbor(MapHex hex, HexSide side)
        {
            return hexTable.Get(hex, side);
        }

        public MapHex GetHex(string locationCode)
        {
            // MapManager is a RailsManager so it is possible to locate by id
            return (MapHex)Locate(locationCode);
        }

        public MapHex.Coordinates Minimum
        {
            get
            {
                return minimum;
            }
        }

        public MapHex.Coordinates Maximum
        {
            get
            {
                return maximum;
            }
        }

        public string MapUIClassName
        {
            get
            {
                return mapOrientation.GetUIClassName();
            }
        }

        public IReadOnlyDictionary<string, StopType> DefaultStopTypes
        {
            get
            {
                return defaultStopTypes;
            }
        }

        public List<Stop> GetCurrentStops()
        {
            List<Stop> stops = new List<Stop>();
            foreach (MapHex hex in hexes.Values)
            {
                stops.AddRange(hex.Stops);
            }
            return stops;
        }

        public List<int> PossibleTileCosts()
        {
            return possibleTileCosts;
        }

        public List<MapHex> ParseLocations(string locationCodes)
        {

            List<MapHex> locationBuilder = new List<MapHex>();
            foreach (string hexName in locationCodes.Split(','))
            {
                MapHex hex = GetHex(hexName);
                if (hex != null)
                {
                    locationBuilder.Add(hex);
                }
                else
                {
                    throw new ArgumentException("Invalid hex " + hexName +
                            " specified in location string " + locationCodes);
                }
            }
            return locationBuilder;
        }

        /**
         * Calculate the distance between two hexes as in 1835,
         * i.e. as "the crow without a passport flies".
         */
        public int GetHexDistance(MapHex hex1, MapHex hex2)
        {
            if (hexDistances == null)
            {
                hexDistances = new Table<MapHex, MapHex, int>();
            }

            if (!hexDistances.Contains(hex1, hex2))
            {
                CalculateHexDistances(hex1, hex1, 0);
            }
            return hexDistances.Get(hex1, hex2);
        }

        private void CalculateHexDistances(MapHex initHex, MapHex currentHex, int depth)
        {
            hexDistances.Put(initHex, currentHex, depth);

            // check for next hexes
            depth++;
            foreach (MapHex nextHex in hexTable.GetRow(currentHex).Values)
            {
                if (!hexDistances.Contains(initHex, nextHex) ||
                        depth < hexDistances.Get(initHex, nextHex))
                {
                    CalculateHexDistances(initHex, nextHex, depth);
                }
            }
        }


        /**
         * Calculate the distances between a given tokenable city hex
         * and all other tokenable city hexes.
         * <p> Distances are cached.
         * @param hex Start hex
         * @return Sorted integer list containing all occurring distances only once.
         */
        public List<int> GetCityDistances(MapHex initHex)
        {
            if (hexDistances == null)
            {
                hexDistances = new Table<MapHex, MapHex, int>();
            }

            if (!hexDistances.ContainsRow(initHex))
            {
                CalculateHexDistances(initHex, initHex, 0);
            }

            List<int> distances = new List<int>();

            foreach (var otherHex in hexDistances.GetRow(initHex))
            {
                if (otherHex.Key.CurrentTile.HasStations)
                {
                    distances.Add(otherHex.Value);
                }
            }
            distances.Sort();
            return distances;
        }

        public string MapImageFilepath
        {
            get
            {
                return mapImageFilepath;
            }
        }

        public int MapXOffset
        {
            get
            {
                return mapXOffset;
            }
            set
            {
                mapXOffset = value;
            }
        }

        public int MapYOffset
        {
            get
            {
                return mapYOffset;
            }
            set
            {
                mapYOffset = value;
            }
        }

        public float MapScale
        {
            get
            {
                return mapScale;
            }
            set
            {
                mapScale = value;
            }
        }

        public bool IsMapImageUsed
        {
            get
            {
                return mapImageUsed;
            }
        }

        public class Table<R, C, V>
        {
            public Table()
            {
                Data = new Dictionary<R, Dictionary<C, V>>();
            }

            public Dictionary<R, Dictionary<C, V>> Data { get; private set; }

            public Dictionary<C, V> GetRow(R row)
            {
                return Data[row];
            }

            public V Get(R row, C col)
            {
                return Data[row][col];
            }

            public void Put(R row, C col, V value)
            {
                if (!ContainsRow(row))
                {
                    Data[row] = new Dictionary<C, V>();
                }
                Data[row][col] = value;
            }

            public bool Contains(R row, C col)
            {
                return Data.ContainsKey(row) && Data[row].ContainsKey(col);
            }

            public bool ContainsRow(R row)
            {
                return Data.ContainsKey(row);
            }
        }
    }
}
